ppcs/ppc0032-optional-strictures.md

```diff
@@ -0,0 +1,176 @@
+under the current `strict` pragma as it would break existing code.
+
+## Rationale
+
+Adding support for optional strictures and warnings provides several benefits:
+
+1. It allows for introducing new strict checks and warnings without breaking
+   backward compatibility
+2. It provides a consistent interface for enabling such checks (`use strict
+   'feature'` rather than `no feature`)
+3. It groups related functionality under the established `strict` and
+   `warnings` pragmas
+4. It provides a clear path for experimental features to graduate into core
+   Perl
+
+The stringification case demonstrates these benefits. Rather than using a
+separate `no stringification` pragma, the functionality can be enabled with
+`use strict 'strings'`, which is more consistent with existing Perl syntax and
+more discoverable through the standard documentation.
+
+## Specification
+
+### Optional Strictures
+
+Add support for optional strict modes that are not enabled by a bare `use
+strict` statement. These would be enabled explicitly:
+
+```perl
+use strict 'strings';  # Enable reference stringification checks
+```
+
+The first such optional stricture would be 'strings', which prevents implicit
+stringification of references. When enabled, the following operations would
+die with an error if attempted on a non-object reference:
+
+- String interpolation (`"$ref"`)
+- String concatenation (`$ref . "foo"`)
+- Case conversion (`lc`, `lcfirst`, `uc`, `ucfirst`)
+- Pattern matching (`$ref =~ m//`)
+- String operations (`split`, `join`)
+- Output operations (`print`, `say`)
+
+Error messages would be specific to the operation being performed, e.g.:
+
+```
+Attempted to stringify a reference at line X
+Attempted to concat a reference at line Y
+```
+
+### Optional Warnings
+
+Add a new warnings classification for optional warnings that are not
+enabled by `use warnings` or `use warnings 'all'`. These would require
+explicit enabling:
+
+```perl
+use warnings 'optional';  # Enable all optional warnings
+use warnings 'optional::specific';  # Enable specific optional warning
+```
+
+## Backwards Compatibility
+
+This proposal is explicitly designed to maintain backward compatibility:
+
+- Existing code using `use strict` or `use warnings` will continue to work
+  unchanged
+- New strictures and warnings must be explicitly enabled
+- The mechanism allows for gradual adoption of new features
+- Existing modules like `stringification` can be updated to use the new
+  mechanism while maintaining their current interface
+
+## Security Implications
+
+The proposed changes improve security by allowing developers to catch more
+potential programming errors. The stringification case in particular can help
+prevent information leaks where reference addresses might be accidentally
+exposed in output, though that seems unlikely to be a significant security
+issue.
+
+## Examples
+
+Here's a trivial example:
+
+```perl
+use strict 'all', 'strings';
```

Comment:
Note there is no "all" flag for strict currently as there is for warnings, so for this example to work, such a flag would also need to be added. But I would suggest a different name (default, standard?), as this proposal is in fact making it not enable all available strictures.

Reply:
D'oh! That's what I get for throwing this together too quickly. I've just pushed a new version.